core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala

```diff
@@ -93,6 +93,29 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
     } {
       out.close()
     }
+
+    val dataFile = getDataFile(shuffleId, mapId)
+    synchronized {
```

Comment:
Can you add a big comment explaining what is going on here?  Also worth noting that there is only one `IndexShuffleBlockResolver` per executor.